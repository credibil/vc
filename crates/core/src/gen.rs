//! # Generate
//!
//! Generate random strinsg for use in authorzation code, token, state,
//! nonce, etc.

// LATER: replace with a proper random string generator

/// Random string generation for auth code, token, state, and nonce.
use base64ct::{Base64UrlUnpadded, Encoding};

const PIN_CHARS: &str = "0123456789";
const PIN_LEN: usize = 6;

// "'`+=,./\|:;?><}{][_-
const SAFE_CHARS: &str =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789)(*&^%$#@!~";
const STATE_LEN: usize = 32;

/// Generates a base64 encoded random string for authorization code.
#[must_use]
pub fn auth_code() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for token
#[must_use]
pub fn token() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for deferred issuance
/// `transaction_id`.
#[must_use]
pub fn transaction_id() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for deferred issuance
/// `transaction_id`.
#[must_use]
pub fn notification_id() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for nonce
#[must_use]
pub fn nonce() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a base64 encoded random string for `issuer_state`
#[must_use]
pub fn state_key() -> String {
    let rnd = random_string(STATE_LEN, SAFE_CHARS);
    Base64UrlUnpadded::encode_string(rnd.as_bytes())
}

/// Generates a use PIN code.
#[must_use]
pub fn tx_code() -> String {
    random_string(PIN_LEN, PIN_CHARS)
}

// Generates a random string from a given set of characters. Uses fastrand so is
// not cryptographically secure.
fn random_string(len: usize, charset: &str) -> String {
    let chars: Vec<char> = charset.chars().collect();
    (0..len).map(|_| chars[fastrand::usize(..chars.len())]).collect()
}

// ///
// /// Generate a new random, base64-encoded 128-bit CSRF token.
// pub fn new_random() -> Self {
//     CsrfToken::new_random_len(16)
// }
// ///
// /// Generate a new random, base64-encoded CSRF token of the specified length.
// ///
// /// # Arguments
// ///
// /// * `num_bytes` - Number of random bytes to generate, prior to
// base64-encoding. pub fn new_random_len(num_bytes: u32) -> Self {
//     let random_bytes: Vec<u8> = (0..num_bytes).map(|_|
// thread_rng().gen::<u8>()).collect();
//     CsrfToken::new(base64::encode_config(&random_bytes,
// base64::URL_SAFE_NO_PAD)) }
