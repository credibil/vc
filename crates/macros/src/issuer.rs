use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::parse::{Error, Result};

use crate::parse::Data;

pub fn create_offer_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let subject_id = fields.remove("subject_id").unwrap();
    let credential_configuration_ids = fields.remove("credential_configuration_ids").unwrap();
    let pre_authorize = fields.remove("pre-authorize").unwrap();
    let tx_code_required = fields.remove("tx_code_required").unwrap();
    let send_type = fields.remove("send_type").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        CreateOfferRequest {
            credential_issuer: #credential_issuer.to_string(),
            subject_id: Some(#subject_id.to_string()),
            credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),
            pre_authorize: #pre_authorize,
            tx_code_required: #tx_code_required,
            send_type:  #send_type,
        }
    })
}

pub fn authorization_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let response_type = fields.remove("response_type").unwrap();
    let client_id = fields.remove("client_id").unwrap();
    let redirect_uri = fields.remove("redirect_uri").unwrap();
    let state = fields.remove("state").unwrap();
    //  let code_challenge = fields.remove("code_challenge").unwrap();
    let code_challenge_method = fields.remove("code_challenge_method").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        AuthorizationRequest {
            credential_issuer: #credential_issuer.to_string(),
            response_type: #response_type.to_string(),
            client_id: #client_id.to_string(),
            redirect_uri: Some(#redirect_uri.to_string()),
            state: Some(#state.to_string()),
            // code_challenge: #code_challenge.to_string(),
            code_challenge_method: #code_challenge_method.to_string(),

            // credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),

            ..Default::default()
        }
    })
}
