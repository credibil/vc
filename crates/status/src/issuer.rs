//! # Status (Issuer)
//!
//! Traits and types for managing the publication of a credential status as an
//! issuer.

use std::future::Future;

pub use vercre_core::Quota;
pub use vercre_w3c_vc::model::CredentialStatus;

use crate::provider;

/// The `Status` trait specifies how status information can be looked up for a
/// credential.
pub trait Status: Send + Sync {
    /// Returns information on how to look up the status of a credential.
    ///
    /// A default implementation is provided that just returns `None` for cases
    /// where the issuer will not be providing a status endpoint.
    fn status(
        &self, _holder_subject: &str, _credential_identifier: &str,
    ) -> impl Future<Output = provider::Result<Option<Quota<CredentialStatus>>>> + Send {
        async { Ok(None) }
    }
}
