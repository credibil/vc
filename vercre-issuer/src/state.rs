//! State is used by the library to persist request information between steps
//! in the issuance process.

use std::collections::HashMap;

use chrono::{DateTime, TimeDelta, Utc};
use serde::{Deserialize, Serialize};
use vercre_openid::issuer::{
    AuthorizationDetail, CodeChallengeMethod, CredentialOffer, CredentialRequest, RequestObject,
};
use vercre_w3c_vc::model::VerifiableCredential;

type CredentialIdentifier = String;

/// State is used to persist request information between issuance steps in the
/// Credential issuance process.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct State {
    /// Identifies the (previously authenticated) Holder in order that Issuer
    /// can authorize credential issuance.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subject_id: Option<String>,

    /// Stage holds data relevant to the current state of the issuance process.
    /// This data is used by subsequent step(s) to verify Wallet interactions,
    /// including credential issuance.
    pub stage: Stage,

    /// Time state should expire.
    pub expires_at: DateTime<Utc>,
}

impl State {
    /// Determines whether state has expired or not.
    pub fn is_expired(&self) -> bool {
        self.expires_at.signed_duration_since(Utc::now()).num_seconds() < 0
    }
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[allow(clippy::large_enum_variant)]
pub enum Stage {
    /// Unauthorized state.
    #[default]
    Unauthorized,

    /// Holds a Credential Offer awaiting retrieval by the Wallet. The Wallet
    /// has been sent a unique URL it can use to retrieve the offer.
    Pending(CredentialOffer),

    /// Holds pre-authorized offer data as presented to the Wallet. This data is
    /// used when validating the Wallet's request for an access token.
    Offered(Offer),

    /// Holds a Pushed Authorization Request awaiting retrieval by the Wallet.
    PushedAuthorization(PushedAuthorization),

    /// Holds authorization data in cases where the Wallet requests and is
    /// granted authorization to request credential issuance. As with
    /// `PreAuthorized` state, this data is used when validating the
    /// Wallet's request for an access token.
    Authorized(Authorization),

    /// Holds information about the access token and corresponding credentials
    /// the Wallet is authorized to request.
    Validated(Token),

    /// Issued Credential state.
    Issued(Credential),

    /// Deferred issuance state.
    Deferred(Deferrance),
}

/// Pre-authorization state from the `create_offer` endpoint.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Offer {
    /// A list of `authorization_details` entries referencing credentials the
    /// Wallet is authorized to request.
    pub items: Option<Vec<AuthorizedItem>>,

    /// Transaction code sent to the holder to use (if present)when requesting
    /// an access token.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tx_code: Option<String>,
}

/// Holds data used during the issuance of a credential.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Authorized {
    /// Identifies the dataset associated with the credential to be issued.
    /// Dataset is unique by issuer not by subject.
    ///
    /// For example, the `credential_configuration_id` is `UniversityDegree_JWT`
    /// and the `credential_identifier` is `EngineeringDegree2023`.
    pub credential_identifier: String,

    /// Credential's `credential_configuration_id` connecting it with supported
    /// credential metadata.
    pub credential_configuration_id: String,

    /// Identifies a subset of claims to use when issuing the associated
    /// credential. This subset is used in cases where the Wallet has
    /// requested (and has been authorized for) issuance of a credential
    /// containing subset of claims.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub claim_ids: Option<Vec<String>>,
}

/// Authorization state.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[allow(clippy::struct_field_names)]
pub struct Authorization {
    /// The `client_id` of the Wallet requesting issuance.
    pub client_id: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub redirect_uri: Option<String>,

    /// PKCE code challenge from the Authorization Request.
    pub code_challenge: String,

    /// PKCE code challenge method from the Authorization Request.
    pub code_challenge_method: CodeChallengeMethod,

    /// A list of authorized `scope` or `authorization_details` entries along
    /// with credential metadata and dataset identifiers.
    pub items: Vec<AuthorizedItem>,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct PushedAuthorization {
    /// The Authorization Request pushed to the PAR endpoint.
    pub request: RequestObject,

    /// The time the request URI should expire at.
    pub expires_at: DateTime<Utc>,
}

/// Authorized `authorization_detail` or `scope` item along with
/// `credential_configuration_id` and `credential_identifier`s.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct AuthorizedItem {
    /// Authorized item.
    #[serde(flatten)]
    pub item: ItemType,

    /// Credential configuration metadata for the item.
    pub credential_configuration_id: String,

    /// Authorized credential datasets for the item.
    pub credential_identifiers: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub enum ItemType {
    /// Authorized item is of type `authorization_detail`
    AuthorizationDetail(AuthorizationDetail),

    /// Authorized item is of type `scope`
    Scope(String),
}

impl Default for ItemType {
    fn default() -> Self {
        Self::AuthorizationDetail(AuthorizationDetail::default())
    }
}

/// Token state.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Token {
    /// The access token.
    #[allow(clippy::struct_field_names)]
    pub access_token: String,

    /// The nonce to be used by the Wallet when creating a proof of possession
    /// of the key proof.
    pub c_nonce: String,

    /// Number denoting the lifetime in seconds of the `c_nonce`.
    pub c_nonce_expires_at: DateTime<Utc>,

    /// Credentials (configuration id and identifier) validated for issuance
    /// using the accompanying access token.
    pub credentials: HashMap<CredentialIdentifier, Authorized>,
}

impl Token {
    pub fn c_nonce_expires_in(&self) -> i64 {
        self.c_nonce_expires_at.signed_duration_since(Utc::now()).num_seconds()
    }

    pub fn c_nonce_expired(&self) -> bool {
        self.c_nonce_expires_in() < 0
    }
}

/// Issued Credential state (for Notification endpoint).
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Credential {
    pub credential: VerifiableCredential,
}

/// Deferred issuance state.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Deferrance {
    /// Used to identify a Deferred Issuance transaction. Is used as the
    /// state persistence key.
    pub transaction_id: String,

    /// Save the Credential request when issuance is deferred.
    pub credential_request: CredentialRequest,
}

pub enum Expire {
    Authorized,
    Access,
    Nonce,
}

impl Expire {
    pub fn duration(&self) -> TimeDelta {
        match self {
            Self::Authorized => TimeDelta::try_minutes(5).unwrap_or_default(),
            Self::Access => TimeDelta::try_minutes(15).unwrap_or_default(),
            Self::Nonce => TimeDelta::try_minutes(10).unwrap_or_default(),
        }
    }
}
