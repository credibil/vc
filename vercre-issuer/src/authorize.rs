//! # Authorization Endpoint
//!
//! The Authorization Endpoint is used by Wallets to request access to the Credential
//! Endpoint, that is, to request issuance of a Credential. The endpoint is used in
//! the same manner as defined in [RFC6749].
//!
//! Wallets can request authorization for issuance of a Credential using
//! `authorization_details` (as defined in [RFC9396]) or `scope` parameters (or both).
//!
//! ## Authorization Requests
//!
//! - One (and only one) of `credential_configuration_id` or `format` is REQUIRED.
//! - `credential_definition` is OPTIONAL.

//! ## Example
//!
//! with `credential_configuration_id`:
//!
//! ```json
//! "authorization_details":[
//!    {
//!       "type": "openid_credential",
//!       "credential_configuration_id": "UniversityDegreeCredential"
//!    }
//! ]
//! ```
//!
//! with `format`:
//!
//! ```json
//! "authorization_details":[
//!     {
//!         "type": "openid_credential",
//!         "format": "vc+sd-jwt",
//!         "vct": "SD_JWT_VC_example_in_OpenID4VCI"
//!     }
//! ]
//! ```
//!
//! **VC Signed as a JWT, Not Using JSON-LD**
//!
//! - `credential_definition` is OPTIONAL.
//!   - `type` is OPTIONAL.
//!   - `credentialSubject` is OPTIONAL.
//!
//! ```json
// ! "authorization_details":[
// !     {
// !         "type": "openid_credential",
// !         "credential_configuration_id": "UniversityDegreeCredential",
// !         "credential_definition": {
// !             "credentialSubject": {
// !                 "given_name": {},
// !                 "family_name": {},
// !                 "degree": {}
// !             }
// !         }
// !     }
// ! ]
//! ```
//! 
//! [RFC6749]: (https://www.rfc-editor.org/rfc/rfc6749.html)
//! [RFC9396]: (https://www.rfc-editor.org/rfc/rfc9396)

// TODO: add support for "ldp_vc" format
// TODO: add support for "jwt_vc_json-ld" format
// TODO: add support for "vc+sd-jwt" format
// LATER: add support for "mso_mdoc" format

use std::collections::HashMap;
use std::fmt::Debug;
use std::vec;

use chrono::Utc;
use tracing::instrument;
use vercre_core::gen;
use vercre_openid::issuer::{
    AuthorizationDetail, AuthorizationDetailType, AuthorizationRequest, AuthorizationResponse,
    AuthorizationSpec, Authorized, ClaimEntry, ConfigurationId, Format, FormatProfile, GrantType,
    Issuer, Metadata, Provider, StateStore, Subject,
};
use vercre_openid::{Error, Result};

// use crate::shell;
use crate::state::{Authorization, Expire, State, Step};

/// Authorization request handler.
///
/// # Errors
///
/// Returns an `OpenID4VP` error if the request is invalid or if the provider is
/// not available.
#[instrument(level = "debug", skip(provider))]
pub async fn authorize(
    provider: impl Provider, request: AuthorizationRequest,
) -> Result<AuthorizationResponse> {
    // get issuer metadata
    let Ok(issuer) = Metadata::issuer(&provider, &request.credential_issuer).await else {
        return Err(Error::InvalidClient("invalid `credential_issuer`".into()));
    };

    let mut ctx = Context {
        issuer,
        ..Context::default()
    };
    ctx.verify(&provider, &request).await?;
    ctx.process(&provider, request).await
}

#[derive(Debug, Default)]
struct Context {
    issuer: Issuer,
    auth_dets: HashMap<String, AuthorizationDetail>,
    scope_items: HashMap<String, String>,
    claims: Option<HashMap<String, ClaimEntry>>,
}

impl Context {
    async fn verify(
        &mut self, provider: &impl Provider, request: &AuthorizationRequest,
    ) -> Result<()> {
        tracing::debug!("authorize::verify");

        // client and server metadata
        let Ok(client) = Metadata::client(provider, &request.client_id).await else {
            return Err(Error::InvalidClient("invalid `client_id`".into()));
        };
        let Ok(server) = Metadata::server(provider, &request.credential_issuer).await else {
            return Err(Error::ServerError("invalid `credential_issuer`".into()));
        };

        // 'authorization_code' grant_type allowed (client and server)?
        let client_grant_types = client.oauth.grant_types.unwrap_or_default();
        if !client_grant_types.contains(&GrantType::AuthorizationCode) {
            return Err(Error::InvalidGrant(
                "authorization_code grant not supported for client".into(),
            ));
        }
        let server_grant_types = server.oauth.grant_types_supported.unwrap_or_default();
        if !server_grant_types.contains(&GrantType::AuthorizationCode) {
            return Err(Error::InvalidRequest(
                "authorization_code grant not supported by server".into(),
            ));
        }

        // is holder identified (authenticated)?
        if request.subject_id.is_empty() {
            return Err(Error::InvalidRequest("missing holder subject".into()));
        }

        // does offer `subject_id`  match request `subject_id`?
        if let Some(issuer_state) = &request.issuer_state {
            let state: State = StateStore::get(provider, issuer_state)
                .await
                .map_err(|e| Error::ServerError(format!("state issue: {e}")))?;

            if state.is_expired() {
                return Err(Error::InvalidRequest("issuer state expired".into()));
            }

            if state.subject_id.as_ref() != Some(&request.subject_id) {
                return Err(Error::InvalidRequest(
                    "request `subject_id` does not match offer".into(),
                ));
            }
        }

        // has a credential been requested?
        if request.authorization_details.is_none() && request.scope.is_none() {
            return Err(Error::InvalidRequest("no credentials requested".into()));
        }

        // verify authorization_details
        if let Some(authorization_details) = &request.authorization_details {
            self.verify_authorization_details(authorization_details)?;
        }
        // verify scope
        if let Some(scope) = &request.scope {
            self.verify_scope(scope)?;
        }

        // redirect_uri
        let Some(redirect_uri) = &request.redirect_uri else {
            return Err(Error::InvalidRequest("no redirect_uri specified".into()));
        };
        let Some(redirect_uris) = client.oauth.redirect_uris else {
            return Err(Error::InvalidRequest("no redirect_uris specified for client".into()));
        };
        if !redirect_uris.contains(redirect_uri) {
            return Err(Error::InvalidRequest("request redirect_uri is not registered".into()));
        }

        // response_type
        if !client.oauth.response_types.unwrap_or_default().contains(&request.response_type) {
            return Err(Error::UnsupportedResponseType(
                "the response_type not supported by client".into(),
            ));
        }
        if !server.oauth.response_types_supported.contains(&request.response_type) {
            return Err(Error::UnsupportedResponseType(
                "response_type not supported by server".into(),
            ));
        }

        // code_challenge
        // N.B. while optional in the spec, we require it
        let challenge_methods = server.oauth.code_challenge_methods_supported.unwrap_or_default();
        if !challenge_methods.contains(&request.code_challenge_method) {
            return Err(Error::InvalidRequest("unsupported code_challenge_method".into()));
        }
        if request.code_challenge.len() < 43 || request.code_challenge.len() > 128 {
            return Err(Error::InvalidRequest(
                "code_challenge must be between 43 and 128 characters".into(),
            ));
        }

        Ok(())
    }

    // Verify Credentials requested in `authorization_details` are supported.
    // N.B. has side effect of saving valid `authorization_detail` objects into context
    // for later use.
    fn verify_authorization_details(
        &mut self, authorization_details: &[AuthorizationDetail],
    ) -> Result<()> {
        let supported = &self.issuer.credential_configurations_supported;

        // check each credential requested is supported by the issuer
        for auth_det in authorization_details {
            if auth_det.type_ != AuthorizationDetailType::OpenIdCredential {
                return Err(Error::InvalidRequest("invalid authorization_details type".into()));
            }

            // verify requested credentials are supported
            match &auth_det.specification {
                AuthorizationSpec::ConfigurationId(ConfigurationId::Definition {
                    credential_configuration_id,
                    credential_definition,
                }) => {
                    //  find supported credential by `credential_configuration_id`
                    if !supported.contains_key(credential_configuration_id) {
                        return Err(Error::InvalidRequest(
                            "unsupported credential_configuration_id".into(),
                        ));
                    }

                    // verify requested claims are supported
                    if let Some(requested_cd) = credential_definition
                        && let Some(claims) = &requested_cd.credential_subject
                    {
                        let supported_cd =
                            &supported[credential_configuration_id].credential_definition;
                        Self::verify_claims(claims, &supported_cd.credential_subject)?;
                        self.claims = Some(claims.clone());
                    }

                    // save `credential_configuration_id` for later use
                    self.auth_dets.insert(credential_configuration_id.clone(), auth_det.clone());
                }
                AuthorizationSpec::Format(Format::JwtVcJson {
                    credential_definition,
                }) => {
                    //  find supported credential by `format` and `type`
                    let Some((config_id, _)) = supported.iter().find(|(_, v)| {
                        v.format == FormatProfile::JwtVcJson
                            && v.credential_definition.type_ == credential_definition.type_
                    }) else {
                        return Err(Error::InvalidRequest(
                            "unsupported credential `format` or `type`".into(),
                        ));
                    };

                    // verify requested claims are supported
                    if let Some(claims) = &credential_definition.credential_subject {
                        let supported_cd = &supported[config_id].credential_definition;
                        Self::verify_claims(claims, &supported_cd.credential_subject)?;
                        self.claims = Some(claims.clone());
                    }

                    // save `credential_configuration_id` for later use
                    self.auth_dets.insert(config_id.clone(), auth_det.clone());
                }
                AuthorizationSpec::ConfigurationId(ConfigurationId::Claims { .. }) => {
                    todo!("ConfigurationId::Claims");
                }
                AuthorizationSpec::Format(Format::LdpVc { .. }) => {
                    todo!("Format::LdpVc");
                }
                AuthorizationSpec::Format(Format::JwtVcJsonLd { .. }) => {
                    todo!("Format::JwtVcJsonLd");
                }
                AuthorizationSpec::Format(Format::MsoDoc { .. }) => {
                    todo!("Format::MsoDoc");
                }
                AuthorizationSpec::Format(Format::VcSdJwt { .. }) => {
                    todo!("Format::VcSdJwt");
                }
            };
        }

        Ok(())
    }

    fn verify_claims(
        requested: &HashMap<String, ClaimEntry>, supported: &Option<HashMap<String, ClaimEntry>>,
    ) -> Result<()> {
        let Some(supp_claims) = &supported else {
            return Ok(());
        };

        for key in requested.keys() {
            if !supp_claims.contains_key(key) {
                return Err(Error::InvalidRequest(format!("{key} claim is not supported")));
            }
        }

        Ok(())
    }

    // Verify Credentials requested in `scope` are supported.
    // N.B. has side effect of saving valid scope items into context for later use.
    fn verify_scope(&mut self, scope: &str) -> Result<()> {
        'verify_scope: for item in scope.split_whitespace() {
            for (config_id, cred_cfg) in &self.issuer.credential_configurations_supported {
                // `authorization_details` credential request  takes precedence `scope` request
                if self.auth_dets.contains_key(config_id) {
                    continue;
                }

                if cred_cfg.scope == Some(item.to_string()) {
                    // save scope item by `credential_configuration_id` for later use
                    self.scope_items.insert(config_id.to_string(), item.to_string());
                    continue 'verify_scope;
                }
            }

            return Err(Error::InvalidRequest("scope item {item} is unsupported".into()));
        }

        Ok(())
    }

    // Authorize Wallet request:
    // - check which requested `credential_configuration_id`s the holder is
    //   authorized for
    // - save related auth_dets/scope items in state
    async fn process(
        &self, provider: &impl Provider, request: AuthorizationRequest,
    ) -> Result<AuthorizationResponse> {
        tracing::debug!("authorize::process");

        // for Credentials requested using `authorization_detail`
        let mut authzd_detail = vec![];
        let mut credentials = HashMap::new();

        for (config_id, auth_det) in &self.auth_dets {
            let identifiers =
                Subject::authorize(provider, &request.subject_id, config_id, self.claims.clone())
                    .await
                    .map_err(|e| Error::ServerError(format!("issue authorizing subject: {e}")))?;

            authzd_detail.push(Authorized {
                authorization_detail: auth_det.clone(),
                credential_identifiers: identifiers.clone(),
            });

            for identifier in &identifiers {
                credentials.insert(identifier.clone(), config_id.clone());
            }
        }

        // for Credentials requested using `scope`
        // FIXME: add `credential_identifiers` to Authorized state
        let mut authzd_scope = vec![];
        for (config_id, scope_item) in &self.scope_items {
            let identifiers = Subject::authorize(provider, &request.subject_id, config_id, None)
                .await
                .map_err(|e| Error::ServerError(format!("issue authorizing holder: {e}")))?;

            authzd_scope.push(scope_item.clone());

            for identifier in &identifiers {
                credentials.insert(identifier.clone(), config_id.clone());
            }
        }

        let authorized = if authzd_detail.is_empty() { None } else { Some(authzd_detail) };

        let scope = if authzd_scope.is_empty() { None } else { Some(authzd_scope.join(" ")) };

        // return an error if holder is not authorized for any requested credentials
        if authorized.is_none() && scope.is_none() {
            return Err(Error::AccessDenied(
                "holder is not authorized for requested credentials".into(),
            ));
        }

        // save authorization state
        let state = State {
            expires_at: Utc::now() + Expire::Authorized.duration(),
            credentials: Some(credentials),
            subject_id: Some(request.subject_id),
            current_step: Step::Authorization(Authorization {
                code_challenge: request.code_challenge,
                code_challenge_method: request.code_challenge_method,
                authorized,
                scope,
                client_id: request.client_id,
                redirect_uri: request.redirect_uri.clone(),
            }),
            ..State::default()
        };

        let code = gen::auth_code();
        StateStore::put(provider, &code, &state, state.expires_at)
            .await
            .map_err(|e| Error::ServerError(format!("state issue: {e}")))?;

        // remove offer state
        if let Some(issuer_state) = &request.issuer_state {
            StateStore::purge(provider, issuer_state)
                .await
                .map_err(|e| Error::ServerError(format!("state issue: {e}")))?;
        }

        Ok(AuthorizationResponse {
            code,
            state: request.state,
            redirect_uri: request.redirect_uri.unwrap_or_default(),
        })
    }
}

#[cfg(test)]
mod tests {
    use base64ct::{Base64UrlUnpadded, Encoding};
    use insta::assert_yaml_snapshot as assert_snapshot;
    use rstest::rstest;
    use sha2::{Digest, Sha256};
    use vercre_macros::authorization_request;
    use vercre_test_utils::issuer::{Provider, CLIENT_ID, CREDENTIAL_ISSUER, NORMAL_USER};
    use vercre_test_utils::snapshot;

    use super::*;

    #[rstest]
    #[case("configuration_id", configuration_id)]
    #[case("credential_definition", credential_definition)]
    #[case("format", format)]
    #[case("scope", scope)]
    async fn request(#[case] name: &str, #[case] f: fn() -> AuthorizationRequest) {
        vercre_test_utils::init_tracer();
        snapshot!("");

        let provider = Provider::new();

        // create request
        let request = f();
        let response = authorize(provider.clone(), request).await.expect("response is ok");

        assert_snapshot!("authorize:configuration_id:response", &response, {
            ".code" => "[code]",
        });

        // compare response with saved state
        let state =
            StateStore::get::<State>(&provider, &response.code).await.expect("state exists");

        assert_snapshot!(format!("authorize:{name}:state"), state, {
            ".expires_at" => "[expires_at]",
            ".current_step.authorized.*.credential_definition.credentialSubject" => insta::sorted_redaction(),
        });
    }

    fn configuration_id() -> AuthorizationRequest {
        authorization_request!({
            "credential_issuer": CREDENTIAL_ISSUER,
            "response_type": "code",
            "client_id": CLIENT_ID,
            "redirect_uri": "http://localhost:3000/callback",
            "state": "1234",
            "code_challenge": Base64UrlUnpadded::encode_string(&Sha256::digest("ABCDEF12345")),
            "code_challenge_method": "S256",
            "authorization_details": [{
                "type": "openid_credential",
                "credential_configuration_id": "EmployeeID_JWT",
            }],
            "subject_id": NORMAL_USER,
            "wallet_issuer": CREDENTIAL_ISSUER
        })
    }

    fn credential_definition() -> AuthorizationRequest {
        authorization_request!({
            "credential_issuer": CREDENTIAL_ISSUER,
            "response_type": "code",
            "client_id": CLIENT_ID,
            "redirect_uri": "http://localhost:3000/callback",
            "state": "1234",
            "code_challenge": Base64UrlUnpadded::encode_string(&Sha256::digest("ABCDEF12345")),
            "code_challenge_method": "S256",
            "authorization_details": [{
                "type": "openid_credential",
                "credential_configuration_id": "EmployeeID_JWT",
                "credential_definition": {
                    "credentialSubject": {
                        "given_name": {},
                        "family_name": {},
                    }
                }
            }],
            "subject_id": NORMAL_USER,
            "wallet_issuer": CREDENTIAL_ISSUER
        })
    }

    fn format() -> AuthorizationRequest {
        authorization_request!({
            "credential_issuer": CREDENTIAL_ISSUER,
            "response_type": "code",
            "client_id": CLIENT_ID,
            "redirect_uri": "http://localhost:3000/callback",
            "state": "1234",
            "code_challenge": Base64UrlUnpadded::encode_string(&Sha256::digest("ABCDEF12345")),
            "code_challenge_method": "S256",
            "authorization_details": [{
                "type": "openid_credential",
                "format": "jwt_vc_json",
                "credential_definition": {
                    "@context": [
                        "https://www.w3.org/2018/credentials/v1",
                        "https://www.w3.org/2018/credentials/examples/v1"
                    ],
                    "type": [
                        "VerifiableCredential",
                        "EmployeeIDCredential"
                    ],
                    "credentialSubject": {}
                }
            }],
            "subject_id": NORMAL_USER,
            "wallet_issuer": CREDENTIAL_ISSUER
        })
    }

    fn scope() -> AuthorizationRequest {
        authorization_request!({
            "credential_issuer": CREDENTIAL_ISSUER,
            "response_type": "code",
            "client_id": CLIENT_ID,
            "redirect_uri": "http://localhost:3000/callback",
            "state": "1234",
            "code_challenge": Base64UrlUnpadded::encode_string(&Sha256::digest("ABCDEF12345")),
            "code_challenge_method": "S256",
            "scope": "EmployeeIDCredential",
            "subject_id": NORMAL_USER,
            "wallet_issuer": CREDENTIAL_ISSUER
        })
    }
}
