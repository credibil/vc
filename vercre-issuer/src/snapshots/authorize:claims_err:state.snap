---
source: vercre-issuer/src/authorize.rs
assertion_line: 414
expression: state
---
subject_id: normal_user
stage:
  authorized:
    client_id: 96bfb9cb-0513-7d64-5532-bed74c48f9ab
    redirect_uri: "http://localhost:3000/callback"
    code_challenge: B6qi8EPWVTnONnO8XNSAsN7O2ejthn3pCKwZUO0HzU8
    code_challenge_method: S256
    items:
      - AuthorizationDetail:
          type: openid_credential
          format: jwt_vc_json
          credential_definition:
            type:
              - VerifiableCredential
              - EmployeeIDCredential
            credentialSubject:
              employee_id: {}
              family_name: {}
              given_name: {}
        credential_configuration_id: EmployeeID_JWT
        credential_identifiers:
          - PHLEmployeeID
expires_at: "[expires_at]"
